keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes a new endpoint to publish the [OpenAPI v3 specification](https://swagger.io/specification/) for Kubernetes types. This solves our problem of stripping fields when publishing the OpenAPI v2 and improves the transparency and accuracy of the published OpenAPI.
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Kubernetes resources and types can be described through their OpenAPI definitions, currently served as OpenAPI v2 at the $cluster/openapi/v2 cluster endpoint. With the introduction of CRDs, users provide an OpenAPI v3 spec to describe the resource. Since kubernetes only publishes OpenAPI v2, these definitions are converted into an OpenAPI v2 definition before being aggregated and served at the /openapi/v2 endpoint. OpenAPI v3 is more expressive than v2, and the conversion results in the loss of some information. Some of the type definitions are just transformed into “accept anything” definitions because of our lack of ability to perform a better conversion and limitations in kubectl. Some efforts at improving the OpenAPI (eg: enum support) are targeted to bo bundled with OpenAPI v3.
+
+Additionally, since kubernetes aggregate the OpenAPI v2 spec for all specs into one endpoint, whenever a spec is modified the entire spec is updated and all clients will need to redownload the spec. Depending on the size and number of CRDs, this is not an inexpensive operation.
+
+### Goals
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Support publishing and aggregating OpenAPI v3 for all kubernetes types
+- Published v3 spec should be a lossless representation of the data and no fields should be stripped for both built-in types and CRDs
+- Instead of serving the entire OpenAPI spec at one endpoint, separate the spec by group and only serve the resources required by each group
```

Comment:
I guess this will come further down: by group is unfortunate as this means that the aggregator has to call out to each APIService and merge specs again. By group-version would remove this need.

Reply:
I don't have any strong opinions about this, and will just list the benefits of each. It does seem like group/version reduces the complexity on the server side while potentially trading some performance on the client side.

Group: 
- Less calls needed to download a subset of the schema
- Total size of schema downloaded is smaller if the entire schema needs to be aggregated client side 

Group/version: 
- APIService does not need any additional aggregation
- CRDs schemas do not need to be merged

@apelisse did a size analysis for group vs group/version

|Split|Cardinality | Avg Size (stddev)| Aggregate Size|
|-- | -- | --| --|
| No split | 1 | 131kB | 131kB |
|Group | 22 | 14kB (12kB)| 308kB|
|Group/Version | 44 | 11kB (10kB)| 484kB|

Reply:
Which clients do you have in mind that care about more or less calls? Is there a client we see that needs the whole spec? I could see some client generators that would, but they don't care about 22 or 44 requests as they are batch-like jobs.

Reply:
I don't know any latency sensitive clients that would care about the actual number of calls, it seems more of a nice to have. Different specs also have overlap that would be reduced by splitting only per group (total aggregate size is 30% smaller), but given the advantages of a omitting an aggregation step with group-version, let's adopt splitting by group-version.